### PR TITLE
Increase code coverage:

### PR DESCRIPTION
**Pull request by prupipho**
## PR checklist

- Quick summary of changes: Adding tests in Core
- Which issue does this PR relate to? #169 

* Extensions tests has been added for:
    * DateTimeExtensions
    * DictionaryExtensions
    * StringExtensions
    * IEnumerableExtensions
    * VersionExtensions
    * TemplateTypeExtensions
 * Helpers tests has been added for:
    * FileHelpers
    * FS (not added coverage for the method **FindFileAtOrAbove**)

* Rename some methods to make them clearer:
    * ObfuscateSHA to be just Obfuscate (used to be several)
    * EnsureFolder to EnsureFolderExists
* Remove some obsolete methods.
* Control some missing scenarios raised by the newly created tests

### CODE REVIEW

**Comment by mrlacey** [comment]
I know this has been added with the title "increase code coverage" but that is all the tests in `DateTimeExtensionsTests.cs` do.
They don't actually test anything. Rather, they repeat the functionality in the code being tested. This is an anti-pattern that doesn't test the functionality but artificially looks to inflate the perceived reliability or quality of the code without doing so. In fact, it risks making the maintenance and support of the codebase harder as tests that copy the implementation are automatically very fragile in their nature without clearly defining the correct behavior.
Hopefully, this will be addressed before this PR comes out of draft.

**Comment by prupipho** [comment]
> cially looks to inflate the perceived reliability or quality of the code without doing so. In fact, it risks making the maintenance and support of the codebase harder as tests that copy the implementation are automatically very fragile in their nature without clearly defining the correct behavior.
> Hopefully, this will be addressed before this PR comes out of d

Sure. We are working on getting them right before completing so they are useful for the purpose of the issue. Thanks for the feedback!

File: code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs [hunk 1]
```diff
@@ -10,11 +10,55 @@ namespace Microsoft.Templates.Core.Test.Extensions
     [Trait("ExecutionSet", "Minimum")]
     public class StringExtensionsTests
     {
+        [Theory]
+        [InlineData("string")]
+        [InlineData("string23TEST")]
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]: Here and below.
The very limited number of test cases is concerning.
Why no tests for whitespace? including at the start, middle, and end?
Why no tests for non-alphanumeric characters?
Why no tests with non-ascii characters?
Why no tests with strings containing non-printing characters?

File: code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs [hunk 2]
```diff
@@ -10,11 +10,55 @@ namespace Microsoft.Templates.Core.Test.Extensions
...
+        public void ObfuscateSHA_ShouldNotBeEmpty(string value)
+        {
+            var actual = value.ObfuscateSHA();
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]: I'm not clear what the method being tested here should do, and these tests don't help.
Does  the `ObfuscateSHA` method obfuscate a string or does it calculate a hash (as implied by the "SHA" part of the name)?
Hashing and obfuscating are not the same thing. Anyone encountering the method in the source code is likely to ask what the method does. The name isn't clear and the tests don't make it clear what the method is supposed to do.
I think some additional documentation (in the code) and possibly a renaming of the method are in order.

File: code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs [hunk 3]
```diff
@@ -10,11 +10,55 @@ namespace Microsoft.Templates.Core.Test.Extensions
...
+
+            Assert.NotEmpty(actual);
+        }
+
+        [Theory]
+        [InlineData("string")]
+        [InlineData("string23TEST")]
+        public void ObfuscateSHA_ShouldNotBeEqualsToTheValue(string value)
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]:
```suggestion
        public void ObfuscateSHA_ShouldNotEqualTheOriginalValue(string value)
```

File: code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs [hunk 4]
```diff
@@ -10,11 +10,55 @@ namespace Microsoft.Templates.Core.Test.Extensions
...
+        {
+            var actual = value.ObfuscateSHA();
+
+            Assert.False(value.Equals(actual, StringComparison.Ordinal));
+        }
+
+        [Theory]
+        [InlineData("string")]
+        [InlineData("string23TEST")]
+        public void ObfuscateSHA_ShouldNotContainLowercase(string value)
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]: This is a really curious test.
Without further information, it's not clear why the method under test should never return a lower case character.
It looks like you might be indirectly testing something tied to the implementation of the method being tested. Does the code that consumes the value returned by the method under test need to have upper case characters.

File: code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs [hunk 5]
```diff
@@ -10,11 +10,55 @@ namespace Microsoft.Templates.Core.Test.Extensions
...
+        {
+            var actual = value.ObfuscateSHA();
+
+            Assert.DoesNotContain(actual, c => char.IsLower(c));
+        }
+
+        [Theory(Skip = "Control null and empty scenarios")]
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]: Is there a follow-up task being created to expand upon this test and address why it is being skipped?
Adding this test raises lots of questions and it's not clear what is meant by "control these scenarios".

Comment (prupipho) [reply code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]: Thanks for all the feedback provided Matt. It´s really helpful. 
I am working on this task to help the team to increase the code coverage in CoreTS, and if possible improving the code itself. I am not too familiar with the code of CoreTemplateStudio myself, that´s why the PR is in draft. This way, I can get feedback from the Team while adding some tests and confirm that the tests are adding some real value in here. 

Depending of what other missing scenarios the code could have, we are going to create a follow-up task or fix those scenarios at the end in this same PR.

File: code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs (hunk 4 above)
Comment (prupipho) [reply code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]: We would change the name of the function to just be **Obfuscate** (the name came from 2017). 
Would it be "**Obfuscate_ShouldHaveBeenConvertedToUppercase**" a better name for this? As the last step of the function is converting the generated string to upper using: ToUpperInvariant.
We think that this method (Obfuscate) should probably be moved to be a private method within Microsoft.Templates.Core.Packaging.TemplatePackage as it´s the only place where it is needed.

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+            var actual = FileHelper.GetFileContent(sourceFile);
+
+            Assert.Empty(actual);
+        }
+
+        [Fact]
+        public void GetFileContent_FileExists_ShouldReturnsCorrectly()
+        {
+            var projectName = "Test";
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs]: as this is only used once, why make it a separate variable?

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            var sourceFile = Path.Combine(Environment.CurrentDirectory, $"TestData\\TestProject\\{projectFile}");
+
+            var actual = FileHelper.GetFileContent(sourceFile);
+
+            Assert.Empty(actual);
+        }
+
+
+        [Fact]
+        public void GetFileContent_Errors_ShouldReturnEmpty()
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs]: This name isn't descriptive of what it does.
"Errors" suggests there will be some sort of error, but none is seen or tested for

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        [InlineData(null)]
+        public void GetFileContent_EmptyFilePath_ShouldReturnEmpty(string sourceFile)
+        {
+            var actual = FileHelper.GetFileContent(sourceFile);
+
+            Assert.Empty(actual);
+        }
+
+        [Theory]
+        [InlineData("TestNotExisting.csproj")]
```
Comment (mrlacey) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs]: Is the plan to add more InlineData input for this method?

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+            Assert.Empty(actual);
+        }
+
+        [Fact]
+        public void GetFileContent_FileExists_ShouldReturnsCorrectly()
+        {
+            var projectName = "Test";
+            var projectFile = $"{projectName}.csproj";
+            var sourceFile = Path.Combine(Environment.CurrentDirectory, $"TestData\\TestProject\\{projectFile}");
```
Comment (prupipho) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs]:
```suggestion
            var sourceFile = Path.Combine(Environment.CurrentDirectory, "TestData\\TestProject\\Test.csproj");
```

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            var actual = FileHelper.GetFileContent(sourceFile);
+
+            Assert.Empty(actual);
+        }
+
+        [Theory]
+        [InlineData("TestNotExisting.csproj")]
+        public void GetFileContent_FileNotExists_ShouldReturnEmpty(string projectFile)
+        {
+            var sourceFile = Path.Combine(Environment.CurrentDirectory, $"TestData\\TestProject\\{projectFile}");
```
Comment (prupipho) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs]:
```suggestion
        [Fact]
        public void GetFileContent_FileNotExists_ShouldReturnEmpty()
        {
            var sourceFile = Path.Combine(Environment.CurrentDirectory, "TestData\\TestProject\\TestNotExisting.csproj");
```

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            var sourceFile = Path.Combine(Environment.CurrentDirectory, "TestData\\TestProject\\TestNotExisting.csproj");
+
+            var actual = FileHelper.GetFileContent(sourceFile);
+
+            Assert.Empty(actual);
+        }
+
+        [Fact]
+        public void GetFileContent_Errors_ShouldReturnEmpty()
```
Comment (prupipho) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs]:
```suggestion
        public void GetFileContent_DirectoryPathRatherThanFilePath_ShouldReturnEmpty()
```

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs (hunk 8 above)
Comment (prupipho) [reply code/test/CoreTemplateStudio.Core.Test/Helpers/FileHelpersTests.cs]: I wanted to test several situations where we can get an error in **File.ReadAllText** but in the end I just added one. Changed this back to [Fact]

**Comment by mrlacey** [comment]
commit 02d6ac0554f44f3846615686e1fc47c34cb920f3 concerns me.
It has the comment "failing pipeline" but all it does is change the order of one of the tests. If the order of the tests can cause them to fail or cause the build pipeline to fail then it hints at bigger issues with the way these tests are being created. 
Does this commit just hide an underlying issue?

**Comment by prupipho** [comment]
> commit [02d6ac0](https://github.com/microsoft/CoreTemplateStudio/commit/02d6ac0554f44f3846615686e1fc47c34cb920f3) concerns me.
> It has the comment "failing pipeline" but all it does is change the order of one of the tests. If the order of the tests can cause them to fail or cause the build pipeline to fail then it hints at bigger issues with the way these tests are being created.
> Does this commit just hide an underlying issue?

It´s just work in progress, I just wanted to make sure that was the root of the problem when I saw the failing pipeline. Thanks.

File: code/test/CoreTemplateStudio.Core.Test/Helpers/FsTests.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            var originalDirectoryExists = Directory.Exists(sourceFolder);
+
+            Fs.EnsureFolder(sourceFolder);
+
+            var newlyCreatedDirectoryExists = Directory.Exists(sourceFolder);
+
+            Assert.False(originalDirectoryExists);
+            Assert.True(newlyCreatedDirectoryExists);
+
+            // TODO HERE Is this the right way to test this and tidy up testing data
```
Comment (prupipho) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/FsTests.cs]: chat with the Team tomorrow to discuss if there´s a nicer approach to this during testing.

File: code/test/CoreTemplateStudio.Core.Test/Extensions/VersionExtensionsTests.cs [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+using System.Collections.Generic;
+using Xunit;
+
+namespace Microsoft.Templates.Core.Test.Extensions
+{
+    [Trait("ExecutionSet", "Minimum")]
+    public class VersionExtensionsTests
+    {
+        [Fact]
+        public void IsZero_IsNullVersion_ShouldReturnTrue()
```
Comment (prupipho) [inline code/test/CoreTemplateStudio.Core.Test/Extensions/VersionExtensionsTests.cs]: rename :)

File: code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Diagnostics/WizardTypeEnum.cs [hunk 14]
```diff
@@ -6,10 +6,10 @@ namespace Microsoft.Templates.Core.Diagnostics
...
-        NewProject = 0,
-        AddPage = 1,
-        AddFeature = 2,
-        AddService = 3,
-        AddTesting = 4,
+        NewProject,
+        AddPage,
+        AddFeature,
+        AddService,
+        AddTesting,
```
Comment (prupipho) [inline code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Diagnostics/WizardTypeEnum.cs]: Best review and change this on a separate issue
```suggestion
        NewProject = 0,
        AddPage = 1,
        AddFeature = 2,
        AddService = 3,
        AddTesting = 4,
```

File: code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs [hunk 15]
```diff
@@ -11,7 +12,7 @@ namespace Microsoft.Templates.Core
 {
     public static class StringExtensions
     {
-        public static string ObfuscateSHA(this string data)
+        public static string Obfuscate(this string data)
```
Comment (prupipho) [inline code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs]: TODO: before closing this PR we should control null scenario

File: code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs [hunk 16]
```diff
@@ -40,12 +41,18 @@ private static string GetHash(HashAlgorithm md5Hash, byte[] inputData)
...
+            if (string.IsNullOrWhiteSpace(value) || string.IsNullOrEmpty(value.Trim()))
             {
                 return new string[0];
             }
 
-            return value.Split("|".ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
+            var values = value.Split("|".ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
+            if (values.Any(v => v != v.Trim()))
+            {
+                throw new InvalidDataException($"Multivalue field: {value} contains trailing or leading whitespaces.");
```
Comment (prupipho) [inline code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs]: Adding a resource for this

File: code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs (hunk 15 above)
Comment (sibille) [reply code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs]: and move this to the TemplatePackage class

File: code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs (hunk 16 above)
Comment (sibille) [reply code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Extensions/StringExtensions.cs]: Nice catch :)

File: code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Helpers/Fs.cs [hunk 17]
```diff
@@ -18,9 +18,16 @@ public static class Fs
...
+            {
+                if (!Directory.Exists(folder))
+                {
+                    Directory.CreateDirectory(folder);
+                }
+            }
+            catch (Exception ex)
             {
-                Directory.CreateDirectory(folder);
+                AppHealth.Current.Error.TrackAsync($"Error creating folder {folder}: {ex.Message}", ex).FireAndForget();
```
Comment (sibille) [inline code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Helpers/Fs.cs]: Saw other methods in the Fs Helper are tracking exceptions as warnings, we should follow the same pattern here

**Comment by prupipho** [comment]
Broken build due to refactoring the FsTests into smaller files and rename done.

File: code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Helpers/Fs.cs [hunk 18]
```diff
@@ -16,65 +14,27 @@ namespace Microsoft.Templates.Core.Helpers
...
-        public static void EnsureFolder(string folder)
+        public static void EnsureFolderExists(string folder)
         {
-            if (!Directory.Exists(folder))
-            {
-                Directory.CreateDirectory(folder);
-            }
-        }
-
-        public static void CopyRecursive(string sourceDir, string targetDir, bool overwrite = false)
```
Comment (sibille) [inline code/src/CoreTemplateStudio/CoreTemplateStudio.Core/Helpers/Fs.cs]: CopyRecursive is used in WinTS in the VisualComparisonTests, we should keep it in the FSHelper for now. #134

File: code/test/CoreTemplateStudio.Core.Test/CoreTemplateStudio.Core.Test.csproj [hunk 19]
```diff
@@ -95,6 +90,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="AutoFixture" Version="4.17.0" />
```
Comment (sibille) [inline code/test/CoreTemplateStudio.Core.Test/CoreTemplateStudio.Core.Test.csproj]: This seems not to be used, we should remove it. Reviewing nuget packages I also saw that Microsoft.NET.Test.Sdk has an update we missed before.

File: code/test/CoreTemplateStudio.Core.Test/Helpers/LogFixture.cs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Text.RegularExpressions;
+using Xunit;
+
+namespace Microsoft.Templates.Core.Test.Helpers.FsTests.Helpers
+{
+    [SuppressMessage("Design", "CA1063:Implement IDisposable Correctly", Justification = "Testing purposes only")]
+    public class LogFixture : IDisposable
```
Comment (sibille) [inline code/test/CoreTemplateStudio.Core.Test/Helpers/LogFixture.cs]: This should probably be called FSTestsFixture as it sets the folderpath and handles log files.

File: code/test/CoreTemplateStudio.Core.Test/CoreTemplateStudio.Core.Test.csproj (hunk 19 above)
Comment (prupipho) [reply code/test/CoreTemplateStudio.Core.Test/CoreTemplateStudio.Core.Test.csproj]: true !! I had a look but didn´t use it in the end!

File: code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs (hunk 1 above)
Comment (prupipho) [reply code/test/CoreTemplateStudio.Core.Test/Extensions/StringExtensionsTests.cs]: Hello @mrlacey !
Thanks a lot for all the feedback provided. We will be merging this PR as it brings some code changes that we need to test in WinTS. If you have any more feedback, we will resolve it on a different Pull Request.

**Review by sibille — APPROVED** [review]

**Pull request merged**